R/disease_progression.R

```diff
@@ -19,30 +19,31 @@ create_infection_update_listener <- function(
   }
 }
 
-#' @title Schedule progression of human disease at the start of the simulation
-#'
-#' @param event the event to schedule
-#' @param state the human state variable
-#' @param from_state the state this event applies to
-#' @param rate the average time spent in this state
-#' @noRd
-initialise_progression <- function(event, state, from_state, rate) {
-  target <- state$get_index_of(from_state)
-  event$schedule(
-    target,
-    log_uniform(target$size(), rate)
-  )
+create_progression_process <- function(
+  events,
+  progression_event,
+  state,
+  from_state,
+  rate
+  ) {
+  function(timestep) {
+    incubating <- events$clinical_infection$get_scheduled()$or(
+      events$asymptomatic_infection$get_scheduled()
+    )
+    progression_event$schedule(
+      state$get_index_of(from_state)$and(incubating$not())$sample(1/rate),
+      0
+    )
```

Comment:
Now disease progression (D -> A -> U -> S) is scheduled in a memoryless way.

Every timestep, we draw an individual from U to transition to S.

We exclude individuals who are incubating. However, one could imagine a model where individuals continue to recover until they are reinfected when their incubation period is over.